src/plugins/dataflow/nodes/utilities/live-output-utilities.ts

```diff
@@ -12,11 +17,52 @@ export function getHubSelect(node: Node) {
 
 export function getOutputType(node: Node) {
   const outputTypeControl = node.controls.get("liveOutputType") as DropdownListControl;
-  return outputTypeControl.getValue();
+  if (outputTypeControl) return outputTypeControl.getValue();
 }
 
 export function getNodeValueWithType(node: Node): NodeOutputValue {
   const val = node.data.nodeValue as number;
   const outType = getOutputType(node);
   return { val, outType };
 }
+
+export function outputsToAnyRelay(node: Node) {
+  return kMicroBitHubRelaysIndexed.includes(getOutputType(node));
+}
+
+export function outputsToAnyGripper(node: Node) {
+  return kGripperOutputTypes.includes(getOutputType(node));
+}
+
+export function getLiveOptions(node: Node, sharedVar?: VariableType, device?: string | null) {
```

Comment:
```suggestion
export function getLiveOptions(node: Node, device?: string | null, sharedVar?: VariableType) {
```
Just to make the argument order the same as `setLiveOutputOpts`.

Reply:
This set off a little chain of events in refactoring and I ended up guaranteeing that `deviceFamily` is a string, by null coalescing it to "unknown device" if its not arduino or microbit,  rather than testing for undefined.  It makes some sense in that that there technically could be a state in which there is a device connected, but we don't know what kind it is (so device family should not be null, just unknown), whereas when there is no serial connection, deviceFamily will legitimately be `undefined`.

Reply:
in dataflow-program `updateNodes`:
```
const foundDeviceFamily = this.stores.serialDevice.deviceFamily ?? "unknown device";
```